### PR TITLE
probe: assess 🤗 repos for potentially malicious files 

### PR DESCRIPTION
**Pull request by leondz**
recreating this one, merged too early, sorry

### CODE REVIEW

File: docs/source/probes.rst [hunk 1]
```diff
@@ -16,6 +16,8 @@ For a detailed oversight into how a probe operates, see :ref:`garak.probes.base.
    garak.probes.dan
    garak.probes.donotanswer
    garak.probes.encoding
+   garak.probes.fileformats
+   garak.probes.gcg
```
Comment (erickgalinkin) [inline docs/source/probes.rst:20]: I believe `garak.probes.gcg` is no longer present since #728 has been merged.

File: garak/detectors/fileformats.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+These detectors examine file formats, based on name or content."""
+
+import pickletools
+
+import magic
+
+from garak.detectors.base import FileDetector
+
+
+class PossiblePickleName(FileDetector):
```
Comment (erickgalinkin) [inline garak/detectors/fileformats.py:15]: Grumble grumble filename-based detections etc etc

File: garak/detectors/fileformats.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    """Guess if filenames look like pickles
+
+    Pickle files have multiple and complex vulnerabilities, and should be
+    avoided. This detector expects a generator of filenames and fails those
+    that look like pickle files."""
+
+    valid_format = None
+
+    def _test_file(self, filename: str) -> None | str:
+        print("Ffffffff", filename)
```
Comment (erickgalinkin) [inline garak/detectors/fileformats.py:25]:
```suggestion
```

File: garak/probes/fileformats.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            return []
+        attempt = self._mint_attempt(None)
+
+        repo_filenames = huggingface_hub.list_repo_files(generator.name)
+        local_filenames = []
+        for repo_filename in tqdm.tqdm(
+            repo_filenames,
+            leave=False,
+            desc=f"Gathering files in {generator.name}",
+            colour="#e5a70e",  # probe theme colour
```
Comment (erickgalinkin) [inline garak/probes/fileformats.py:61]: Should we consider storing the colors somewhere like `garak.resources.common` to reuse them more easily?

File: tests/detectors/test_detectors.py [hunk 5]
```diff
@@ -58,26 +60,41 @@ def test_detector_detect(classname):
...
+            4,
+        ), "detector should return as many results as in all_outputs (maybe excluding Nones)"
+
+
+def test_filedetector_nonexist():
+    d = garak.detectors.base.FileDetector()
+    a = garak.attempt.Attempt(prompt="")
+    a.outputs = [None, "", "/non/existing/file"]
+    a.notes["format"] = d.valid_format
+    assert list(d.detect(a)) == []
```
Comment (erickgalinkin) [inline tests/detectors/test_detectors.py:100]: Should we have a message for this assert?

Comment (erickgalinkin) [reply tests/detectors/test_detectors.py:100]: Also, if we want this to be an empty list, we should probably check `len` since this might miss some weirdo outliers where `list(d.detect(a))` is some other value where the test returns true.

File: tests/resources/fileformats/exec_files/grep-mach-o-top4k
Comment (erickgalinkin) [inline tests/resources/fileformats/exec_files/grep-mach-o-top4k:1]: I feel kinda yucky about shipping binaries of any kind as part of the package. Not sure how else we'd test it, but I'm not happy about it.

**Review by erickgalinkin — CHANGES_REQUESTED** [review]
Worth discussing this in more detail. Also, how does this compare/differ from #754

File: tests/resources/fileformats/exec_files/grep-mach-o-top4k
Comment (jmartin-tech) [reply tests/resources/fileformats/exec_files/grep-mach-o-top4k:1]: I theory `test` files should not be shipped as part of the package, they are here as a fixture for the tests and seem valid as things that come with a developer `git` based install.

File: garak/detectors/fileformats.py (hunk 2 above)
Comment (leondz) [reply garak/detectors/fileformats.py:15]: Would you rather we skipped this entirely? I'm not married to it

File: tests/resources/fileformats/exec_files/grep-mach-o-top4k
Comment (leondz) [reply tests/resources/fileformats/exec_files/grep-mach-o-top4k:1]: I should say these are all broken, just the first 4k of the executable. Will add a filter for tests in pip packaging. How could they further be "defanged"? Magic docs say it wants at least 2K

**Comment by leondz** [comment]
@erickgalinkin  no difference with #754 - i merged that by mistake, reverted, rebuilt the PR (which is this) and it got a new id

File: tests/resources/fileformats/exec_files/grep-mach-o-top4k
Comment (leondz) [reply tests/resources/fileformats/exec_files/grep-mach-o-top4k:1]: `test/` dir excluded by [MANIFEST.in](https://github.com/leondz/garak/blob/main/MANIFEST.in), & just verified w/ a pip install

Comment (leondz) [reply tests/resources/fileformats/exec_files/grep-mach-o-top4k:1]: truncated bins also now stored as base64, only decoded during running tests

File: garak/probes/fileformats.py (hunk 4 above)
Comment (leondz) [reply garak/probes/fileformats.py]: now in `garak.resources.theme` 😊

File: tests/detectors/test_detectors_fileformats.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+
+def test_fileisexectuable_exec():
+    exec_attempt = garak.attempt.Attempt(prompt="")
+    exec_attempt.notes["format"] = "local filename"
+    path_prefix = Path("tests/resources/fileformats/exec_files/")
+
+    encoded_exec_filenames = glob.glob(str(path_prefix / "*.base64"))
+    exec_filenames = []
+    for encoded_exec_filename in encoded_exec_filenames:
```
Comment (jmartin-tech) [inline tests/detectors/test_detectors_fileformats.py:118]: The decoding process here may be better implemented as a `fixture` that registers the file to be deleted post test, as is a failure of the `assert` would leave all decoded files on disk in the temporary location.

Untested possible refactor:

``` Python
path_prefix = Path("tests/resources/fileformats/exec_files/")

@pytest.fixture
def decoded_file(request, encoded_exec_filename):
    with open(encoded_exec_filename, "r", encoding="utf-8") as encodedfile:
        with tempfile.NamedTemporaryFile(mode="wb", delete=False) as binfile:
            binfile.write(binascii.a2b_base64(encodedfile.read()))
            binfile.close()
            exec_filenames.append(binfile.name)

    def remove_decoded():
        with contextlib.suppress(FileNotFoundError):
            os.remove(binfile.name)

    return binfile.name
```
``` Python
@pytest.mark.parametrize("encoded_exec_filename", glob.glob(str(path_prefix / "*.base64")))
def test_fileisexectuable_exec(decoded_file):
    d = garak.detectors.fileformats.FileIsExecutable()
    exec_attempt.outputs = list(decoded_file)
    results = d.detect(exec_attempt)
    assert list(results) == [1.0]
```

Comment (leondz) [reply tests/detectors/test_detectors_fileformats.py]: yup, implemented

File: tests/detectors/test_detectors_fileformats.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    def remove_decoded():
+        with contextlib.suppress(FileNotFoundError):
+            Path(binfile.name).unlink()
+
+    request.addfinalizer(remove_decoded)
+
+
+@pytest.mark.parametrize(
+    "encoded_exec_filename", glob.glob(str(path_prefix / "*.base64"))
+)
```
Comment (jmartin-tech) [inline tests/detectors/test_detectors_fileformats.py:133]: May be a good idea to skip these test when `magic` is not available, not all development environments will need to interact with this probe, contributors will be more likely to run tests if they don't need to hit edge cases to get the suite to run at all.

```suggestion
try:
    import magic
except ImportError as e:
    magic == None


@pytest.mark.skipif(magic is None, reason="libmagic not available")
@pytest.mark.parametrize(
    "encoded_exec_filename", glob.glob(str(path_prefix / "*.base64"))
)
```

File: garak/detectors/fileformats.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import pickletools
+
+try:
+    import magic
+except ImportError as e:
+    logging.info(
+        "detectors.fileformats: couldn't import python-magic, try installing libmagic, e.g. `brew install libmagic`",
+        e,
+    )
+    magic == None
```
Comment (jmartin-tech) [inline garak/detectors/fileformats.py:18]:
```suggestion
    magic = None
```

``` Python
>>> magic == None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'magic' is not defined
```

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
One minor syntax issue, and some recommendations.

File: tests/detectors/test_detectors_fileformats.py (hunk 7 above)
Comment (leondz) [reply tests/detectors/test_detectors_fileformats.py:133]: yes, this makes sense, good catch

File: garak/detectors/fileformats.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (leondz) [reply garak/detectors/fileformats.py]: yup, will do, good point

File: pyproject.toml [hunk 9]
```diff
@@ -67,6 +67,9 @@ dependencies = [
   "fschat>=0.2.36",
   "litellm>=1.33.8",
   "jsonpath-ng>=1.6.1",
+  "huggingface_hub>=0.21.0",
+  'python-magic-bin; sys_platform == "win32"',
+  'python-magic>=0.4.21; sys_platform != "win32"',
```
Comment (jmartin-tech) [inline pyproject.toml:72]: Sorry late thought, should we set a version restriction here?

Comment (jmartin-tech) [reply pyproject.toml:72]:
```suggestion
  'python-magic-bin>=0.4.14; sys_platform == "win32"',
  'python-magic>=0.4.21; sys_platform != "win32"',
```

File: requirements.txt [hunk 10]
```diff
@@ -28,6 +28,9 @@ deepl==1.17.0
 fschat>=0.2.36
 litellm>=1.33.8
 jsonpath-ng>=1.6.1
+huggingface_hub>=0.21.0
+python-magic-bin; sys_platform == "win32"
```
Comment (jmartin-tech) [inline requirements.txt:32]:
```suggestion
python-magic-bin>=0.4.14; sys_platform == "win32"
```
